app-framework/include/host/enclave_platform.h

```diff
@@ -288,12 +288,110 @@ namespace monza::host
       if (
         aligned_shmem_offset >= shmem_offset &&
         aligned_shmem_offset + size > aligned_shmem_offset &&
-        aligned_shmem_offset + size < QEMU_SHMEM_SIZE)
+        aligned_shmem_offset + size < SHMEM_SIZE)
+      {
+        memset(shmem_base + aligned_shmem_offset, 0, size);
+        auto result = std::make_pair(
+          shmem_base + aligned_shmem_offset,
+          SHMEM_START + aligned_shmem_offset);
+        shmem_offset = aligned_shmem_offset + size;
+        return result;
+      }
+      else
+      {
+        throw std::runtime_error(
+          "No enough enclave shared memory for allocation.");
+      }
+    }
+
+    friend class EnclavePlatform<InitializerTuple>;
+  };
+
+#endif
+
+#ifdef MONZA_HOST_SUPPORTS_HCS
+
+  template<typename InitializerTuple>
+  class HcsEnclavePlatform : public EnclavePlatform<InitializerTuple>
+  {
+    static constexpr size_t SHMEM_SIZE = 64 * 1024 * 1024;
+
+    std::unique_ptr<HCSEnclaveAbstract> instance;
+
+    size_t shmem_guest_base;
+    uint8_t* shmem_base;
+    size_t shmem_offset;
+
+    bool joined;
+
+  protected:
+    HcsEnclavePlatform(
+      EnclaveType type, const std::string& path, size_t num_threads)
+    : EnclavePlatform<InitializerTuple>(num_threads),
+      instance(HCSEnclaveAbstract::create(path, num_threads, SHMEM_SIZE)),
+      shmem_guest_base(instance->shared_memory_guest_base()),
+      shmem_base(instance->shared_memory().data()),
+      shmem_offset(0),
+      joined(false)
+    {
+      // Reserve space for InitializerTuple.
+      if (SHMEM_SIZE > sizeof(InitializerTuple))
```

Comment:
Could `InitializerTuple` have an alignment requirement?

Reply:
The start of the region would be at least 4kB, but more than likely 2MB aligned. And the InitializerTuple is placed at the very start. So any sensible alignment would be easily met, but it could break for some very specific large alignments.